src/coprocessor/cache.rs

```diff
@@ -0,0 +1,490 @@
+    result: Vec<u8>,
+    size: usize,
+}
+
+impl DistSQLCacheEntry {
+    pub fn new(region_id: u64, version: u64, key_size: usize, r: Vec<u8>) -> DistSQLCacheEntry {
+        let size = r.len() + DISTSQL_CACHE_ENTRY_ADDITION_SIZE + 2 * key_size;
+        DistSQLCacheEntry {
+            region_id: region_id,
+            version: version,
+            result: r,
+            size: size,
+        }
+    }
+}
+
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
+    cached_items: HashMap<DistSQLCacheKey, u8>,
+}
+
+
+#[derive(Default)]
+pub struct DistSQLCache {
+    regions: HashMap<u64, RegionDistSQLCacheEntry>,
+    max_size: usize,
+    map: LinkedHashMap<DistSQLCacheKey, Box<DistSQLCacheEntry>>,
+    size: usize,
+}
+
+impl DistSQLCache {
+    // capacity is memory size unit is byte
+    pub fn new(capacity: usize) -> DistSQLCache {
+        DistSQLCache {
+            max_size: capacity,
+            ..Default::default()
+        }
+    }
+
+    pub fn put(&mut self, region_id: u64, k: DistSQLCacheKey, version: u64, res: Vec<u8>) {
+        if !self.is_region_cache_enabled(region_id) {
+            return;
+        }
+
+        let key_size = k.len();
+        if self.map.contains_key(&k) {
+            let mut entry = self.map.get_mut(&k).unwrap();
+            let old_size = entry.size;
+            entry.size = res.len() + (key_size * 2) + DISTSQL_CACHE_ENTRY_ADDITION_SIZE;
+            entry.version = version;
+            entry.result = res;
+            entry.region_id = region_id;
+            self.size = self.size - old_size + entry.size;
+        } else {
+            let entry = box DistSQLCacheEntry::new(region_id, version, key_size, res);
+            self.size += entry.size;
+            self.map.insert(k.clone(), entry);
+            self.update_regions(region_id, k);
+        }
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    fn check_evict_key(&mut self, region_id: u64, k: &str) {
+        let opt = match self.map.get(k) {
+            None => None,
+            Some(entry) => {
+                match self.regions.get(&region_id) {
+                    None => None,
+                    Some(rentry) => {
+                        // Region's version is not same as cached evict it
+                        if rentry.version != entry.version {
+                            Some(())
+                        } else {
+                            None
+                        }
+                    }
+                }
+            }
+        };
+        if opt.is_some() {
+            self.remove(k);
+        }
+    }
```

Comment:
Could we use the following implement to make it more easy(from L104 to L122)?

```
        match (self.map.get(k), self.regions.get(&region_id)) {
            (Some(entry), Some(rentry)) => if entry.version == rentry.version {
                return;
            },
            _ => return,
        };
        self.remove(k);
```

Reply:
@AndreMouche I think this will calculate `self.regions.get(...)` always even if we don't need.